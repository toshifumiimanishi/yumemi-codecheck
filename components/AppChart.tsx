import { useContext, useState, useEffect } from 'react'
import Highcharts from 'highcharts'
import HighchartsReact from 'highcharts-react-official'
import AppContext from '../contexts/AppContext'

type Props = {
  options: {}
}

const AppChart: React.FC<Props> = ({ options }) => (
  <HighchartsReact highcharts={Highcharts} options={options} />
)

const ContainerAppChart: React.FC = () => {
  const { state } = useContext(AppContext)
  const [options, setOptions] = useState({
    chart: {
      type: 'spline',
    },
    title: {
      text: '都道府県別の総人口推移グラフ',
    },
    legend: {
      layout: 'vertical',
      align: 'right',
      verticalAlign: 'middle',
    },
    xAxis: {
      title: {
        text: '年度（年）',
      },
      categories: [],
    },
    yAxis: {
      title: {
        text: '人口数（人）',
      },
    },
    series: [],
  })

  const initializeHighcharts = () => {
    const newOptions = { ...options }
    const { xAxis } = newOptions

    Highcharts.setOptions({
      lang: {
        numericSymbols: ['万', '億'],
        numericSymbolMagnitude: 10000,
      },
    })

    state.totalPopulation.map((population) => {
      const years = population.data.map(({ year }) => year)

      xAxis.categories = years
      setOptions(newOptions)
    })
  }

  const updateSeries = () => {
    const newOptions = { ...options }
    const newSeries = state.totalPopulation.map((population) => {
      const data = population.data.map(({ value }) => value)

      return {
        name: population.prefName,
        data,
      }
    })

    newOptions.series = newSeries
    setOptions(newOptions)
  }

  useEffect(() => {
    updateSeries()
  }, [state.totalPopulation])

  useEffect(() => {
    initializeHighcharts()
  }, [])

  return <AppChart options={options} />
}

export default ContainerAppChart
