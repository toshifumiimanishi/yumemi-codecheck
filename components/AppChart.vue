<template>
  <highcharts class="chart" :options="chartOptions"></highcharts>
</template>

<script>
import Highcharts from 'highcharts'
import { Chart } from 'highcharts-vue'

Highcharts.setOptions({
  lang: {
    numericSymbols: ['万', '億'],
    numericSymbolMagnitude: 10000
  },
});

export default {
  props: {
    totalPopulation: {
      type: Array
    }
  },

  components: {
    highcharts: Chart
  },

  data() {
    return {
      chartOptions: {
        chart: {
          type: 'spline'
        },
        title: {
          text: '都道府県別の総人口推移グラフ'
        },
        legend: {
          layout: 'vertical',
          align: 'right',
          verticalAlign: 'middle'
        },
        xAxis: {
          title: {
            text: '年度（年）',
          },
          categories: []
        },
        yAxis: {
          title: {
            text: '人口数（人）'
          }
        },
        series: []
      }
    };
  },

  methods: {
    updateCategoriesOnce() {
      const { xAxis } = this.chartOptions;
      const isUpdated = xAxis.categories.length > 0;

      if (!isUpdated) {
        const totalPopulation = this.totalPopulation;

        totalPopulation.map(population => {
          const years = population.data.map(({ year }) => year);

          xAxis.categories = years;
        });
      }
    },

    updateSeries() {
      const totalPopulation = this.totalPopulation;
      const newSeries = totalPopulation.map(population => {
        const data = population.data.map(({ value }) => value);

        return {
          name: population.prefName,
          data
        }
      })

      this.chartOptions.series = newSeries;
    }
  },

  watch: {
    totalPopulation() {
      this.updateCategoriesOnce();
      this.updateSeries();
    }
  }
};
</script>

<style>
.chart {
  margin-top: 36px;
}
</style>
