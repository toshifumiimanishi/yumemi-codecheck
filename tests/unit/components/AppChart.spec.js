import { mount } from '@vue/test-utils';
import AppChart from '~/components/AppChart.vue';
import totalPopulation from '../_mockData/totalPopulation.json';

describe('AppChart コンポーネントテスト', () => {
  const propsData = {
    totalPopulation
  };

  it('レンダリング結果のスナップショットテスト', () => {
    const wrapper = mount(AppChart, { propsData });
    expect(wrapper.vm.$el).toMatchSnapshot();
  })
});
