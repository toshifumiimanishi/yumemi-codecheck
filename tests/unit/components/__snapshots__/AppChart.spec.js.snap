// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppChart コンポーネントテスト レンダリング結果のスナップショットテスト 1`] = `
<div
  class="chart"
  data-highcharts-chart="0"
  style="overflow: hidden;"
>
  <div
    class="highcharts-container "
    dir="ltr"
    id="highcharts-8me7o7q-0"
    style="position: relative; overflow: hidden; width: 600px; height: 400px; text-align: left; line-height: normal; z-index: 0; user-select: none;"
  >
    <svg
      class="highcharts-root"
      height="400"
      style="font-family:\\"Lucida Grande\\", \\"Lucida Sans Unicode\\", Arial, Helvetica, sans-serif;font-size:12px;"
      version="1.1"
      viewBox="0 0 600 400"
      width="600"
      xmlns="http://www.w3.org/2000/svg"
    >
      <desc>
        Created with Highcharts 8.1.2
      </desc>
      <defs>
        <clippath
          id="highcharts-8me7o7q-1-"
        >
          <rect
            fill="none"
            height="375"
            width="580"
            x="0"
            y="0"
          />
        </clippath>
      </defs>
      <rect
        class="highcharts-background"
        fill="#ffffff"
        height="400"
        rx="0"
        ry="0"
        width="600"
        x="0"
        y="0"
      />
      <rect
        class="highcharts-plot-background"
        fill="none"
        height="375"
        width="580"
        x="10"
        y="10"
      />
      <rect
        class="highcharts-plot-border"
        data-z-index="1"
        fill="none"
        height="375"
        width="580"
        x="10"
        y="10"
      />
      <g
        class="highcharts-series-group"
        data-z-index="3"
      />
      <text
        class="highcharts-title"
        data-z-index="4"
        style="color:#333333;font-size:18px;fill:#333333;"
        text-anchor="middle"
        x="300"
        y="24"
      >
        <tspan>
          都道府県別の総人口推移グラフ
        </tspan>
      </text>
      <text
        class="highcharts-subtitle"
        data-z-index="4"
        style="color:#666666;fill:#666666;"
        text-anchor="middle"
        x="300"
        y="10"
      />
      <text
        class="highcharts-caption"
        data-z-index="4"
        style="color:#666666;fill:#666666;"
        text-anchor="start"
        x="10"
        y="397"
      />
      <g
        class="highcharts-legend"
        data-z-index="7"
      >
        <rect
          class="highcharts-legend-box"
          fill="none"
          height="8"
          rx="0"
          ry="0"
          visibility="hidden"
          width="8"
          x="0"
          y="0"
        />
        <g
          data-z-index="1"
        >
          <g />
        </g>
      </g>
      <text
        class="highcharts-credits"
        data-z-index="8"
        style="cursor:pointer;color:#999999;font-size:9px;fill:#999999;"
        text-anchor="end"
        x="590"
        y="395"
      >
        Highcharts.com
      </text>
    </svg>
  </div>
</div>
`;
